src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocGenerator.java

```diff
@@ -0,0 +1,68 @@
+// checkstyle: Checks Java source code and other text files for adherence to a set of rules.
+// Copyright (C) 2001-2023 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.internal.utils;
+
+import java.io.File;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.util.Set;
+
+import org.apache.maven.doxia.parser.Parser;
+import org.apache.maven.doxia.sink.Sink;
+import org.apache.maven.doxia.sink.SinkFactory;
+import org.codehaus.plexus.DefaultPlexusContainer;
+import org.codehaus.plexus.PlexusContainer;
+import org.codehaus.plexus.util.ReaderFactory;
+
+import com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser;
+import com.puppycrawl.tools.checkstyle.site.XdocsTemplateSinkFactory;
+
+public final class XdocGenerator {
+    private static final String XDOCS_TEMPLATE_HINT = "xdocs-template";
+
+    private XdocGenerator() {
+    }
+
+    public static void generateXdocContent() throws Exception {
+        final PlexusContainer plexus = new DefaultPlexusContainer();
+        final Set<Path> templatesFilePaths = XdocUtil.getXdocsTemplatesFilePaths();
+
+        for (Path path : templatesFilePaths) {
+            final String pathToFile = path.toString();
+            final File inputFile = new File(pathToFile);
+            final File outputFile = new File(pathToFile.replace(".template", ""));
+
+            final XdocsTemplateSinkFactory sinkFactory = (XdocsTemplateSinkFactory)
+                    plexus.lookup(SinkFactory.ROLE, XDOCS_TEMPLATE_HINT);
+            final Sink sink = sinkFactory.createSink(outputFile.getParentFile(),
+                    outputFile.getName(), String.valueOf(StandardCharsets.UTF_8));
+            final XdocsTemplateParser parser = (XdocsTemplateParser)
+                    plexus.lookup(Parser.ROLE, XDOCS_TEMPLATE_HINT);
+            try (Reader reader = ReaderFactory.newReader(inputFile,
+                    String.valueOf(StandardCharsets.UTF_8))) {
+                parser.parse(reader, sink);
+            }
+            finally {
+                sink.close();
```

Comment:
https://github.com/apache/maven-doxia/blob/24ab8ba193ec44007333775576277f9aa9ea165f/doxia-sink-api/src/main/java/org/apache/maven/doxia/sink/Sink.java#L55

can we use 
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html ?

Reply:
Unfortunately no. 

> Incompatible types. Found: 'org.apache.maven.doxia.sink.Sink', required: 'java.lang.AutoCloseable'

```java
            try (Reader reader = ReaderFactory.newReader(inputFile,
                    String.valueOf(StandardCharsets.UTF_8));
                Sink sink = sinkFactory.createSink(outputFile.getParentFile(),
                    outputFile.getName(), String.valueOf(StandardCharsets.UTF_8))) {
                parser.parse(reader, sink);
            }
```

Reply:
Intellij says `Sink` is not `AutoCloseable`

> Incompatible types. Found: 'org.apache.maven.doxia.sink.Sink', required: 'java.lang.AutoCloseable'

It implemented `AutoCloseable` 4 months ago in [doxia-2.0.0-M7](https://github.com/apache/maven-doxia/releases/tag/doxia-2.0.0-M7).

I am not sure if this version is stable or what the M7 means.

https://github.com/apache/maven-doxia/commit/b113dc75a5982bb503e0e7c1be5da815744ef166

Reply:
ok, no problem, let not update to most recent for now.